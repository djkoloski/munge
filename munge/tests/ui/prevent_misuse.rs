extern crate core;
extern crate munge;

use {
    ::core::mem::MaybeUninit,
    ::munge::munge,
};

fn main() {
    struct Example {
        a: u32,
        b: u32,
    }

    let mut mu = MaybeUninit::<Example>::uninit();

    munge!(let Example { a, b } = &mut mu);
    assert_eq!(a.write(1), &1);
    assert_eq!(b.write(2), &2);

    // SAFETY: `mu` is completely initialized.
    let value = unsafe { mu.assume_init() };
    //^ ERROR: cannot move out of `mu` because it is borrowed
    //^ NOTE: move out of `mu` occurs here
    assert_eq!(value.a, 1);
    assert_eq!(value.b, 2);

    a.write(3);
    //^ NOTE: borrow later used here
}
