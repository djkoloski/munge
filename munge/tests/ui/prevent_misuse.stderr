error[E0505]: cannot move out of `mu` because it is borrowed
  --> $DIR/prevent_misuse.rs:22:26
   |
17 |     munge!(let Example { a, b } = &mut mu);
   |                                   ------- borrow of `mu` occurs here
...
22 |     let value = unsafe { mu.assume_init() };
   |                          ^^ move out of `mu` occurs here
...
28 |     a.write(3);
   |     ---------- borrow later used here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0505`.
